crates/pallet-domains/src/lib.rs

```diff
@@ -368,6 +425,62 @@ mod pallet {
             Ok(())
         }
 
+        #[pallet::call_index(5)]
+        #[pallet::weight(Weight::from_all(10_000))]
+        // TODO: proper benchmark
+        pub fn submit_bundle_v2(
+            origin: OriginFor<T>,
+            opaque_bundle: OpaqueBundleOf<T>,
+        ) -> DispatchResult {
+            ensure_none(origin)?;
+
+            log::trace!(target: "runtime::domains", "Processing bundle: {opaque_bundle:?}");
+
+            let domain_id = opaque_bundle.domain_id();
+            let bundle_hash = opaque_bundle.hash();
+            let extrinsics_root = opaque_bundle.extrinsics_root();
+            let operator_id = opaque_bundle.operator_id();
+            let (bundle_author, receipt) = opaque_bundle.deconstruct();
+
+            let receipt_type = execution_receipt_type::<T>(domain_id, &receipt);
+            if receipt_type == ReceiptType::Stale {
+                // The stale receipt should not be further processed, but we still track them for purposes
+                // of measuring the bundle production rate.
+                Self::note_domain_bundle(domain_id);
+                return Ok(());
+            }
+
+            // Add the exeuctione receipt to the block tree
+            process_execution_receipt::<T>(domain_id, operator_id, receipt, receipt_type)
+                .map_err(Error::<T>::from)?;
+
+            // `SuccessfulBundles` is empty means this is the first accepted bundle for this domain in this
+            // consensus block, which also mean a domain block will be produced thus update `HeadDomainNumber`
+            // to this domain block's block number.
+            if SuccessfulBundles::<T>::get(domain_id).is_empty() {
+                let next_number = HeadDomainNumber::<T>::get(domain_id)
+                    .checked_add(&One::one())
+                    .ok_or::<Error<T>>(BlockTreeError::MaxHeadDomainNumber.into())?;
+                HeadDomainNumber::<T>::set(domain_id, next_number);
+            }
+
+            // Put the `extrinsics_root` to the inbox of the current under building domain block
+            let head_domain_number = HeadDomainNumber::<T>::get(domain_id);
+            ExecutionInbox::<T>::append(domain_id, head_domain_number, extrinsics_root);
```

Comment:
Bundle does not have the notion of fork like ER as it does the notion of parent bundle.

In the block tree there are forks because the ER or the resulting state root after execution is unknown by the consensus chain because the bundle is executed by the operator. But here the consensus chain can know exactly which bundle is accepted in which consensus block and for a consensus block that contains bundles a domain block must be produced by the honest operator.

Reply:
Ahh make sense Thanks!